DynamicActiveQuery.php

```diff
@@ -28,6 +30,43 @@ class DynamicActiveQuery extends ActiveQuery
     private $dynamicColumn;
 
     /**
+     * Convert index value to closure, that will get decoded dynamic attribute, in case if indexing attribute is dynamic
+     * @param callable|string $column
+     * @return $this
+     */
+    public function indexBy($column)
+    {
+        if (!$this->asArray) {
+            return parent::indexBy($column);
+        }
+
+        /** @var \yii\db\ActiveRecord $modelClass */
+        $modelClass = $this->modelClass;
+        $this->indexBy = function ($row) use ($column, $modelClass) {
+            if (isset($row[$column])) {
+                return $row[$column];
+            }
+
+            if (!method_exists($modelClass, 'dynamicColumn')) {
+                throw new UnknownMethodException("Method {$modelClass}::dynamicColumn() does not exist");
+            }
+
+            $dynamicColumn = $modelClass::dynamicColumn();
+            if (!isset($row[$dynamicColumn])) {
+                throw new UnknownPropertyException("Dynamic column {$dynamicColumn} does not exist");
+            }
```

Comment:
I added a test for that - `DynamicActiveQueryTest::testIndexByExceptionForMissingDynColumn()`